### PR TITLE
DPE-2040 scale down refactor

### PR DESCRIPTION
**Pull request by paulomach**
## Issue

Scale down is not using proper tear down events.

## Solution

Use `storage_detaching` event for tearing down units (as done for VM).

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -396,37 +397,6 @@ def _join_unit_to_cluster(self) -> None:
...
-            return
-
-        if not addresses_of_units_to_remove:
-            return
-
-        self.unit.status = MaintenanceStatus("Removing scaled down units from cluster")
-
-        for unit_address in addresses_of_units_to_remove:
-            try:
-                self._mysql.force_remove_unit_from_cluster(unit_address)
```
Comment (shayancanonical) [inline src/charm.py:424]: we should delete the `force_remove_unit_from_cluster` method as well

**Review by taurus-forever — APPROVED** [review]

File: src/relations/mysql_provider.py [hunk 2]
```diff
@@ -292,6 +292,11 @@ def _on_database_broken(self, event: RelationBrokenEvent) -> None:
             # run once by the leader
             return
 
+        if self.charm.unit_peer_data.get("unit-status", None) == "removing":
+            # safeguard against relation broken being triggered for
+            # a unit being torn down (instead of un-related)
+            return
```
[the remark this answers is not in the document]
Comment (paulomach) [reply src/relations/mysql_provider.py:298]: None that I'm aware

File: src/charm.py (hunk 1 above)
Comment (paulomach) [reply src/charm.py:424]: c28ae64

File: src/relations/mysql_provider.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (paulomach) [reply src/relations/mysql_provider.py:298]: just tested a couple times and storage_detaching is *before* any relation broken/departed event

**Review by shayancanonical — APPROVED** [review]

**Pull request merged**